File: frontend/src/api/appPackages.ts
import { z } from "zod";
import { apiFetch } from "./client";

const DesignerPropFieldSchema = z.object({
  key: z.string(),
  labelKey: z.string(),
  placeholder: z.string().optional(),
  default: z.string().optional(),
});

const DesignerManifestSchema = z.object({
  catalogId: z.string(),
  serviceType: z.string(),
  components: z.array(
    z.object({
      id: z.string(),
      label: z.string(),
      role: z.string(),
    }),
  ),
  chain: z.array(
    z.object({
      from: z.string(),
      to: z.string(),
      label: z.string().optional(),
    }),
  ),
  roles: z.record(
    z.string(),
    z.object({
      props: z.array(DesignerPropFieldSchema),
    }),
  ),
  hydrate: z.enum(["none", "onDrop", "poll"]).optional(),
  detailPathTemplate: z.string().optional(),
  applySteps: z
    .object({
      createUnbound: z
        .object({ hrefTemplate: z.string(), messageKey: z.string() })
        .optional(),
      openBound: z
        .object({ hrefTemplate: z.string(), messageKey: z.string() })
        .optional(),
      componentExtras: z
        .array(
          z.object({
            whenRole: z.string(),
            hrefTemplate: z.string(),
            messageKey: z.string(),
          }),
        )
        .optional(),
    })
    .optional(),
});

export type ApiDesignerManifest = z.infer<typeof DesignerManifestSchema>;

const AppPackageSummarySchema = z.object({
  id: z.string(),
  serviceType: z.string(),
  version: z.string(),
  name: z.string(),
  summary: z.string(),
  reference: z.boolean().optional(),
  hydrate: z.enum(["none", "onDrop", "poll"]).optional(),
  actions: z.array(z.string()).optional(),
});

export type AppPackageSummary = z.infer<typeof AppPackageSummarySchema>;

const AppPackageCatalogCardSchema = z.object({
  id: z.string(),
  serviceType: z.string(),
  version: z.string(),
  reference: z.boolean().optional(),
  name: z.string(),
  summary: z.string(),
  description: z.string(),
  kind: z.string(),
  category: z.string(),
  brand: z.record(z.string(), z.unknown()),
  tags: z.array(z.string()).default([]),
  capabilities: z.array(z.string()).default([]),
  primaryAction: z.string().nullable().optional(),
  implementationHint: z.string().nullable().optional(),
  notes: z.array(z.string()).default([]),
  flowNodes: z
    .array(
      z.object({
        id: z.string(),
        label: z.string(),
        role: z.string(),
      }),
    )
    .default([]),
  flowEdges: z
    .array(
      z.object({
        from: z.string(),
        to: z.string(),
        label: z.string().optional(),
      }),
    )
    .default([]),
});

export type AppPackageCatalogCard = z.infer<typeof AppPackageCatalogCardSchema>;

export function fetchDesignerManifests(options?: {
  includeReference?: boolean;
}): Promise<ApiDesignerManifest[]> {
  const params = new URLSearchParams();
  if (options?.includeReference) params.set("includeReference", "true");
  const qs = params.toString();
  return apiFetch(
    `/api/v1/app-packages/designer-manifests${qs ? `?${qs}` : ""}`,
    (data) => z.array(DesignerManifestSchema).parse(data),
  );
}

export function fetchAppPackages(options?: {
  includeReference?: boolean;
}): Promise<AppPackageSummary[]> {
  const params = new URLSearchParams();
  if (options?.includeReference) params.set("includeReference", "true");
  const qs = params.toString();
  return apiFetch(`/api/v1/app-packages${qs ? `?${qs}` : ""}`, (data) =>
    z.array(AppPackageSummarySchema).parse(data),
  );
}

export function fetchAppPackageCatalog(options?: {
  includeReference?: boolean;
}): Promise<AppPackageCatalogCard[]> {
  const params = new URLSearchParams();
  if (options?.includeReference) params.set("includeReference", "true");
  const qs = params.toString();
  return apiFetch(`/api/v1/app-packages/catalog${qs ? `?${qs}` : ""}`, (data) =>
    z.array(AppPackageCatalogCardSchema).parse(data),
  );
}

const AppStorePackageSchema = z.object({
  id: z.string(),
  version: z.string(),
  name: z.string(),
  summary: z.string(),
  source: z.enum(["bundled", "github"]),
  path: z.string().optional().nullable(),
  archiveUrl: z.string().optional().nullable(),
  signatureUrl: z.string().optional().nullable(),
  repository: z.string().optional().nullable(),
  storeId: z.string().optional().nullable(),
  storeName: z.string().optional().nullable(),
  installed: z.boolean(),
  installedVersion: z.string().optional().nullable(),
  signing: z.enum(["required", "not_applicable", "signed", "unsigned"]).optional().nullable(),
});

const AppStoreSourceMetaSchema = z.object({
  id: z.string(),
  name: z.string(),
  indexUrl: z.string().nullable().optional(),
  indexSource: z.enum(["remote", "bundled"]).optional().default("bundled"),
  priority: z.number().optional().default(0),
});

const AppStoreIndexSchema = z.object({
  apiVersion: z.string(),
  name: z.string(),
  indexSource: z.enum(["remote", "bundled"]).optional().default("bundled"),
  indexUrl: z.string().nullable().optional(),
  sources: z.array(AppStoreSourceMetaSchema).optional().default([]),
  packages: z.array(AppStorePackageSchema),
});

export type AppStoreIndex = z.infer<typeof AppStoreIndexSchema>;
export type AppStorePackage = z.infer<typeof AppStorePackageSchema>;

const AppPackageInstallResultSchema = z.object({
  id: z.string(),
  version: z.string(),
  status: z.enum(["installed", "already_installed"]),
});

export type AppPackageInstallResult = z.infer<typeof AppPackageInstallResultSchema>;

export function fetchAppStore(options?: {
  includeReference?: boolean;
}): Promise<AppStoreIndex> {
  const params = new URLSearchParams();
  if (options?.includeReference) params.set("includeReference", "true");
  const qs = params.toString();
  return apiFetch(`/api/v1/app-packages/store${qs ? `?${qs}` : ""}`, (data) =>
    AppStoreIndexSchema.parse(data),
  );
}

export function installAppPackage(payload: {
  packageId?: string;
  archiveUrl?: string;
  signatureUrl?: string;
}): Promise<AppPackageInstallResult> {
  return apiFetch(
    "/api/v1/app-packages/install",
    (data) => AppPackageInstallResultSchema.parse(data),
    {
      method: "POST",
      body: payload,
    },
  );
}

const AppStoreSourceSchema = z.object({
  id: z.string(),
  name: z.string(),
  indexUrl: z.string(),
  enabled: z.boolean(),
  priority: z.number(),
});

const AppStoreTrustKeySchema = z.object({
  id: z.string(),
  name: z.string(),
  publicKey: z.string(),
});

const AppStoreTrustSchema = z.object({
  allowUnsignedPackages: z.boolean(),
  keys: z.array(AppStoreTrustKeySchema),
});

export type AppStoreSource = z.infer<typeof AppStoreSourceSchema>;
export type AppStoreTrust = z.infer<typeof AppStoreTrustSchema>;
export type AppStoreTrustKey = z.infer<typeof AppStoreTrustKeySchema>;

export function fetchAppStoreSources(): Promise<AppStoreSource[]> {
  return apiFetch("/api/v1/app-store/sources", (data) =>
    z.array(AppStoreSourceSchema).parse(data),
  );
}

export function replaceAppStoreSources(sources: AppStoreSource[]): Promise<AppStoreSource[]> {
  return apiFetch("/api/v1/app-store/sources", (data) => z.array(AppStoreSourceSchema).parse(data), {
    method: "PUT",
    body: { sources },
  });
}

export function createAppStoreSource(payload: {
  id?: string;
  name: string;
  indexUrl: string;
  enabled?: boolean;
  priority?: number;
}): Promise<AppStoreSource> {
  return apiFetch("/api/v1/app-store/sources", (data) => AppStoreSourceSchema.parse(data), {
    method: "POST",
    body: payload,
  });
}

export function deleteAppStoreSource(sourceId: string): Promise<void> {
  return apiFetch(`/api/v1/app-store/sources/${encodeURIComponent(sourceId)}`, () => undefined, {
    method: "DELETE",
  });
}

export function fetchAppStoreTrust(): Promise<AppStoreTrust> {
  return apiFetch("/api/v1/app-store/trust", (data) => AppStoreTrustSchema.parse(data));
}

export function updateAppStoreTrust(payload: {
  allowUnsignedPackages: boolean;
  keys?: AppStoreTrustKey[];
}): Promise<AppStoreTrust> {
  return apiFetch("/api/v1/app-store/trust", (data) => AppStoreTrustSchema.parse(data), {
    method: "PUT",
    body: payload,
  });
}

export function createAppStoreTrustKey(payload: {
  id?: string;
  name: string;
  publicKey: string;
}): Promise<AppStoreTrustKey> {
  return apiFetch("/api/v1/app-store/trust/keys", (data) => AppStoreTrustKeySchema.parse(data), {
    method: "POST",
    body: payload,
  });
}

export function deleteAppStoreTrustKey(keyId: string): Promise<void> {
  return apiFetch(`/api/v1/app-store/trust/keys/${encodeURIComponent(keyId)}`, () => undefined, {
    method: "DELETE",
  });
}

File: backend/app/schemas/app_packages.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class AppPackageSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    serviceType: str
    version: str
    name: str
    summary: str
    reference: bool = False
    hydrate: Literal["none", "onDrop", "poll"] = "none"
    actions: list[str] = Field(default_factory=list)


class AppPackageRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    serviceType: str
    version: str
    reference: bool = False
    capabilities: dict[str, Any]
    catalog: dict[str, Any]
    designer: dict[str, Any]
    instanceSchema: dict[str, Any]
    desiredStateExample: dict[str, Any]
    flowNodes: list[dict[str, Any]] = Field(default_factory=list)
    flowEdges: list[dict[str, Any]] = Field(default_factory=list)


class DesignerManifestRead(BaseModel):
    """Mirrors frontend DesignerManifest (package designer.json)."""

    model_config = ConfigDict(extra="allow")

    catalogId: str
    serviceType: str
    components: list[dict[str, Any]]
    chain: list[dict[str, Any]]
    roles: dict[str, Any]
    hydrate: Literal["none", "onDrop", "poll"] | None = None
    detailPathTemplate: str | None = None
    applySteps: dict[str, Any] | None = None


class AppPackageCatalogCard(BaseModel):
    """Catalog-ready overlay card derived from package catalog.json + designer flow."""

    model_config = ConfigDict(extra="forbid")

    id: str
    serviceType: str
    version: str
    reference: bool = False
    name: str
    summary: str
    description: str
    kind: str
    category: str
    brand: dict[str, Any]
    tags: list[str] = Field(default_factory=list)
    capabilities: list[str] = Field(default_factory=list)
    primaryAction: str | None = None
    implementationHint: str | None = None
    notes: list[str] = Field(default_factory=list)
    flowNodes: list[dict[str, Any]] = Field(default_factory=list)
    flowEdges: list[dict[str, Any]] = Field(default_factory=list)


class AppStorePackage(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    version: str
    name: str
    summary: str
    source: Literal["bundled", "github"]
    path: str | None = None
    archiveUrl: str | None = None
    signatureUrl: str | None = None
    repository: str | None = None
    storeId: str | None = None
    storeName: str | None = None
    installed: bool = False
    installedVersion: str | None = None
    signing: Literal["required", "not_applicable", "signed", "unsigned"] | None = None


class AppStoreSourceMeta(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    indexUrl: str | None = None
    indexSource: Literal["remote", "bundled"] = "bundled"
    priority: int = 0


class AppStoreIndexRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    apiVersion: str
    name: str
    indexSource: Literal["remote", "bundled"] = "bundled"
    indexUrl: str | None = None
    sources: list[AppStoreSourceMeta] = Field(default_factory=list)
    packages: list[AppStorePackage] = Field(default_factory=list)


class AppPackageInstallRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    packageId: str | None = None
    archiveUrl: str | None = None
    signatureUrl: str | None = None


class AppPackageInstallResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    version: str
    status: Literal["installed", "already_installed"]


class AppStoreSource(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    indexUrl: str
    enabled: bool = True
    priority: int = 0


class AppStoreSourcesReplace(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sources: list[AppStoreSource] = Field(default_factory=list)


class AppStoreSourceCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str | None = None
    name: str
    indexUrl: str
    enabled: bool = True
    priority: int = 0


class AppStoreTrustKey(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    publicKey: str


class AppStoreTrustRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    allowUnsignedPackages: bool = True
    keys: list[AppStoreTrustKey] = Field(default_factory=list)


class AppStoreTrustUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    allowUnsignedPackages: bool
    keys: list[AppStoreTrustKey] | None = None


class AppStoreTrustKeyCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str | None = None
    name: str
    publicKey: str
